### ISSUE TITLE
[Bug] bfloat16 下模型测试结果 远远差于fp16/fp32

### ISSUE THREAD

**Issue by mmmans** [InternLM/InternLM#221]
### Describe the bug

用如下三种方法load 模型，并测试发现bfloat16测试掉点严重，请问是有其他细节要注意吗
```
model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.bfloat16)

model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.float16)

model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.float32)
```
perplexity 对比

![image](https://github.com/InternLM/InternLM/assets/24487097/e9d583eb-d830-46d5-aa92-91594eef0f8b)

benchmark 上测试bfloata16也是比官方 report低很多, fp32 基本能对上官方测试结果

另外我看 hf 上 config.json 标的 torch_dtype 就是 bfloat16, ..............
### Environment

transformers == 4.30.2

**Comment by yhcc** [comment]
感谢反馈，我们这边也排查一下~

**Comment by daxia-daxia** [comment]
any progress?  I found that the logits fo internlm-7b is extremely large, which is 570+ for my case. This would get rounded in bfloat16 model. Is this the root cause?

**Comment by scenarios** [comment]
Any progress?

**Comment by daxia-daxia** [comment]
Maybe you guys train internlm-7b using deepspeed zero-1 ? @yhcc

**Comment by scenarios** [comment]
> 感谢反馈，我们这边也排查一下~

Do you have any suggestions on this issue?

**Comment by scenarios** [comment]
> ### Describe the bug
> 用如下三种方法load 模型，并测试发现bfloat16测试掉点严重，请问是有其他细节要注意吗
> 
> ```
> model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.bfloat16)
> 
> model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.float16)
> 
> model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.float32)
> ```
> 
> perplexity 对比
> 
> ![image](https://user-images.githubusercontent.com/24487097/262592409-e9d583eb-d830-46d5-aa92-91594eef0f8b.png)
> 
> benchmark 上测试bfloata16也是比官方 report低很多, fp32 基本能对上官方测试结果
> 
> 另外我看 hf 上 config.json 标的 torch_dtype 就是 bfloat16, ..............
> 
> ### Environment
> transformers == 4.30.2
> 
> ### Other information
> _No response_

Have you tested float16? Does it lead to performance degeneration too?

**Comment by mmmans** [comment]
> > ### Describe the bug
> > 用如下三种方法load 模型，并测试发现bfloat16测试掉点严重，请问是有其他细节要注意吗
> > ```
> > model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.bfloat16)
> > 
> > model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.float16)
> > 
> > model = AutoModelForCausalLM.from_pretrained(ckpt_path, trust_remote_code=True, torch_dtype=torch.float32)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > perplexity 对比
> > ![image](https://user-images.githubusercontent.com/24487097/262592409-e9d583eb-d830-46d5-aa92-91594eef0f8b.png)
> > benchmark 上测试bfloata16也是比官方 report低很多, fp32 基本能对上官方测试结果
> > 另外我看 hf 上 config.json 标的 torch_dtype 就是 bfloat16, ..............
> > ### Environment
> > transformers == 4.30.2
> > ### Other information
> > _No response_
> 
> Have you tested float16? Does it lead to performance degeneration too?

no, only test float32 and bfloat16.
